# Cleaning an *eBay Kleinanzeige* dataset. 

The dataset contains used car data with 50,000 data points. Using pandas to explore the data.

*    dateCrawled - When the ad was first crawled. All field-values are taken from this date.
*    name - Name of the car.
*    seller - Whether the seller is private or a dealer.
*    offerType - The type of listing.
*    price - The listed selling price of the car.
*    abtest - Whether the listing is included in an A/B test.
*    vehicleType - The type of vehicle.
*    yearOfRegistration - The year in which the car was first registered.
*    gearbox - The type of transmission.
*    powerPS - The power of the car in PS.
*    model - The car model name.
*    odometer - How many kilometers the car has driven.
*    monthOfRegistration - The month in which the car was first registered.
*    fuelType - What type of fuel the car uses.
*    brand - The brand of the car.
*    notRepairedDamage - If the car has a damage which is not yet repaired.
*    dateCreated - The date the eBay listing was created.
*    nrOfPictures - The number of pictures in the ad.
*    postalCode - The postal code for the location of the vehicle.
*    lastSeenOnline - When the crawler saw this ad last online.


In [2]:
import numpy as np
import pandas as pd

In [3]:
autos = pd.read_csv("autos.csv", encoding = "latin1")

In [5]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [6]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [7]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [8]:
autos.shape

(50000, 20)

Most of the columns have strings, while the others have numeric values. Preliminary exploration suggests that there are null or missing values in a few of the columns. 

In [9]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
autos.rename({'yearOfRegistration':'registration_year',
             'monthOfRegistration':'registration_month',
             'notRepairedDamage':'unrepaired_damage',
             'dateCreated':'ad_created',
             'dateCrawled':'date_crawled',
             'offerType':'offer_type',
             'abtest':'ab_test',
             'vehicleType':'vehicle_type',
             'powerPS':'power_PS',
             'fuelType':'fuel_type',
             'nrOfPictures':'num_of_pictures',
             'postalCode':'postal_code',
             'lastSeen':'last_seen'}, axis=1, inplace=True)

In [11]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                0   
1         bmw              nein  2016-04-04 00:00:00                0   
2  volkswagen              nein  2016-03-26 00:00:00                0   
3       smart              nein  2016-03-12 00:00:00                0   
4        ford              nein  2016-04-01 00:00:00                0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Column names were changed to match Python conventions more closely.

In [13]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_of_pictures   postal_code  \
count                 50000          50000.0  50000.000000   
unique                   76              NaN           NaN   
top     2016-04-03 00:00:00              NaN           NaN   
freq                   1946              NaN           NaN   
mean                    NaN              0.0  50813.627300   
std                     NaN              0.0  25779.747957   
min                     NaN              0.0   1067.000000   
25%                     NaN              0.0  30451.000000   
50%                     NaN              0.0  49577.000000   
75%                     NaN              0.0  71540.000000   
max                     NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [14]:
autos["seller"].unique()

array(['privat', 'gewerblich'], dtype=object)

In [15]:
autos["offer_type"].unique()

array(['Angebot', 'Gesuch'], dtype=object)

In [16]:
autos["ab_test"].unique()

array(['control', 'test'], dtype=object)

In [17]:
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [19]:
autos["num_of_pictures"].unique()

array([0])

In [22]:
autos["price"].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

A number of columns can be removed since these contain minimal/repetitive information, as can be seen from the unique counts of those columns. The 'num_of_pictures' column contain no information at all. Preserving 'gearbox', since knowing what kind of transmission a car has got is important.

The 'price' column is converted to numeric, so it the odometer column.

In [5]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","").astype(float)
autos.rename({"price":"price_in_$"}, axis=1, inplace=True)

/tmp/ipykernel_64/916684939.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$","")


In [26]:
autos["price_in_$"].dtype

dtype('float64')

In [6]:
autos["odometer"] = autos["odometer"].str.replace("[,km]","").astype(float)
autos.rename({"odometer":"odometer_in_km"}, axis=1, inplace=True)

/tmp/ipykernel_64/1628371182.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  autos["odometer"] = autos["odometer"].str.replace("[,km]","").astype(float)


In [29]:
autos["odometer_in_km"].dtype

dtype('float64')

In [31]:
dropped_cols = ["seller", "offer_type", "ab_test", "num_of_pictures"]
autos.drop(dropped_cols, axis=1, inplace=True)


In [32]:
autos.columns

Index(['date_crawled', 'name', 'price_in_$', 'vehicle_type',
       'registration_year', 'gearbox', 'power_PS', 'model', 'odometer_in_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [33]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        50000 non-null  object 
 1   name                50000 non-null  object 
 2   price_in_$          50000 non-null  float64
 3   vehicle_type        44905 non-null  object 
 4   registration_year   50000 non-null  int64  
 5   gearbox             47320 non-null  object 
 6   power_PS            50000 non-null  int64  
 7   model               47242 non-null  object 
 8   odometer_in_km      50000 non-null  float64
 9   registration_month  50000 non-null  int64  
 10  fuel_type           45518 non-null  object 
 11  brand               50000 non-null  object 
 12  unrepaired_damage   40171 non-null  object 
 13  ad_created          50000 non-null  object 
 14  postal_code         50000 non-null  int64  
 15  last_seen           50000 non-null  object 
dtypes: f

Looking at the missing values...

In [8]:
autos["vehicle_type"].value_counts(dropna=False)

limousine     12859
kleinwagen    10822
kombi          9127
NaN            5095
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

In [40]:
autos[autos["vehicle_type"].isnull()].iloc[0]

date_crawled          2016-03-22 16:51:34
name                           Seat_Arosa
price_in_$                          250.0
vehicle_type                          NaN
registration_year                    2000
gearbox                           manuell
power_PS                                0
model                               arosa
odometer_in_km                   150000.0
registration_month                     10
fuel_type                             NaN
brand                                seat
unrepaired_damage                    nein
ad_created            2016-03-22 00:00:00
postal_code                          7426
last_seen             2016-03-26 18:18:10
Name: 8, dtype: object

In [42]:
autos["name"].value_counts()

Ford_Fiesta                                                 78
Volkswagen_Golf_1.4                                         75
BMW_316i                                                    75
BMW_318i                                                    72
Volkswagen_Polo                                             72
                                                            ..
peugeot_206_bj_2000rot_4tuerig_ohne_tuev_bastlerfahrzeug     1
Volvo_Volvo_S60_2.4                                          1
Sparsamer_Mitsubishi_Cold                                    1
SEAT_ALTEA_2.0_TDI__AUTOMATIK                                1
Fiat_500_C_1.2_Dualogic_Lounge                               1
Name: name, Length: 38754, dtype: int64

In [9]:
autos[autos["gearbox"].isnull()].iloc[0]

date_crawled                 2016-03-16 18:45:34
name                  Mercedes_A140_Motorschaden
seller                                    privat
offer_type                               Angebot
price_in_$                                 350.0
ab_test                                  control
vehicle_type                                 NaN
registration_year                           2000
gearbox                                      NaN
power_PS                                       0
model                                        NaN
odometer_in_km                          150000.0
registration_month                             0
fuel_type                                 benzin
brand                              mercedes_benz
unrepaired_damage                            NaN
ad_created                   2016-03-16 00:00:00
num_of_pictures                                0
postal_code                                17498
last_seen                    2016-03-16 18:45:34
Name: 11, dtype: obj

Exploring the values in odometer_in_km and price_in_$ columns to see if there are outliers.

In [47]:
autos["price_in_$"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_in_$, dtype: float64

In [49]:
autos["price_in_$"].isnull().sum()

0

In [60]:
autos["price_in_$"].value_counts().sort_index(ascending=False)

99999999.0       1
27322222.0       1
12345678.0       3
11111111.0       2
10000000.0       1
              ... 
5.0              2
3.0              1
2.0              3
1.0            156
0.0           1421
Name: price_in_$, Length: 2357, dtype: int64

In [67]:
autos["price_in_$"].value_counts().sort_index(ascending=False).index[1]

27322222.0

In [78]:
autos.loc[autos["price_in_$"]<500,"price_in_$"].value_counts().sort_index(ascending=False)

499.0      92
495.0       4
490.0      34
485.0       2
480.0      25
         ... 
5.0         2
3.0         1
2.0         3
1.0       156
0.0      1421
Name: price_in_$, Length: 139, dtype: int64

In [83]:
autos.loc[autos["price_in_$"]>50000,"price_in_$"].value_counts().sort_index(ascending=False)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
             ..
51900.0       2
51500.0       1
51000.0       1
50900.0       1
50500.0       2
Name: price_in_$, Length: 151, dtype: int64

In [70]:
autos[autos["price_in_$"]==autos["price_in_$"].value_counts()
      .sort_index(ascending=False).index[1]]

date_crawled               name  price_in_$ vehicle_type  \
42221  2016-03-08 20:39:05  Leasinguebernahme  27322222.0    limousine   

       registration_year  gearbox  power_PS model  odometer_in_km  \
42221               2014  manuell       163    c4         40000.0   

       registration_month fuel_type    brand unrepaired_damage  \
42221                   2    diesel  citroen               NaN   

                ad_created  postal_code            last_seen  
42221  2016-03-08 00:00:00        76532  2016-03-08 20:39:05

In [71]:
autos[autos["price_in_$"]==autos["price_in_$"].value_counts()
      .sort_index(ascending=False).index[0]]

date_crawled                         name  price_in_$  \
39705  2016-03-22 14:58:27  Tausch_gegen_gleichwertiges  99999999.0   

      vehicle_type  registration_year    gearbox  power_PS     model  \
39705    limousine               1999  automatik       224  s_klasse   

       odometer_in_km  registration_month fuel_type          brand  \
39705        150000.0                   9    benzin  mercedes_benz   

      unrepaired_damage           ad_created  postal_code            last_seen  
39705               NaN  2016-03-22 00:00:00        73525  2016-04-06 05:15:30

Exploring the prices I've decided to denote prices `> $100000 and prices < $500` to be outliers. There are ~1400 cars which are listed for $0. 

In [7]:
autos_trim = autos[autos["price_in_$"].between(500,100000)]

In [86]:
autos_trim["price_in_$"].describe()

count    45058.000000
mean      6177.407342
std       7636.932471
min        500.000000
25%       1500.000000
50%       3500.000000
75%       7900.000000
max      99900.000000
Name: price_in_$, dtype: float64

In [87]:
autos_trim["price_in_$"].value_counts().sort_index(ascending=False)

99900.0      2
99000.0      2
98500.0      1
94999.0      1
93911.0      1
          ... 
520.0        8
517.0        1
510.0        2
501.0        1
500.0      781
Name: price_in_$, Length: 2171, dtype: int64

In [88]:
autos["odometer_in_km"].value_counts().sort_index(ascending=False)

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
40000.0       819
30000.0       789
20000.0       784
10000.0       264
5000.0        967
Name: odometer_in_km, dtype: int64

Nothing to be done with the odometer values

In [13]:
autos_trim.shape

(45058, 20)

---
Checking out the 5 columns with dates. ```date_crawled, last_seen, and ad_created``` are strings, whereas ```registration_month and registration_year``` are numeric.

Starting with ```date_crawled```:

In [11]:
autos_trim["date_crawled"].str[:10]

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 45058, dtype: object

In [15]:
autos_trim["date_crawled"].str[:10].value_counts(
    normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.025545
2016-03-06    0.014137
2016-03-07    0.036220
2016-03-08    0.033157
2016-03-09    0.032913
2016-03-10    0.032736
2016-03-11    0.033046
2016-03-12    0.037330
2016-03-13    0.015536
2016-03-14    0.036331
2016-03-15    0.034001
2016-03-16    0.029362
2016-03-17    0.031160
2016-03-18    0.012850
2016-03-19    0.034777
2016-03-20    0.038084
2016-03-21    0.037729
2016-03-22    0.032935
2016-03-23    0.032403
2016-03-24    0.028985
2016-03-25    0.031093
2016-03-26    0.032625
2016-03-27    0.031160
2016-03-28    0.034822
2016-03-29    0.033290
2016-03-30    0.033290
2016-03-31    0.031670
2016-04-01    0.033867
2016-04-02    0.035798
2016-04-03    0.038817
2016-04-04    0.036642
2016-04-05    0.013161
2016-04-06    0.003174
2016-04-07    0.001354
Name: date_crawled, dtype: float64

```ad_created```:

In [16]:
autos_trim["ad_created"].str[:10].value_counts(
    normalize=True, dropna=False).sort_index(ascending=True)

2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
                ...   
2016-04-03    0.039039
2016-04-04    0.037019
2016-04-05    0.011896
2016-04-06    0.003262
2016-04-07    0.001198
Name: ad_created, Length: 76, dtype: float64

```last_seen```:


In [17]:
autos_trim["last_seen"].str[:10].value_counts(
    normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.001087
2016-03-06    0.004172
2016-03-07    0.005216
2016-03-08    0.007013
2016-03-09    0.009454
2016-03-10    0.010298
2016-03-11    0.012051
2016-03-12    0.023925
2016-03-13    0.008877
2016-03-14    0.012295
2016-03-15    0.015691
2016-03-16    0.016157
2016-03-17    0.027698
2016-03-18    0.007390
2016-03-19    0.015425
2016-03-20    0.020418
2016-03-21    0.020662
2016-03-22    0.021261
2016-03-23    0.018399
2016-03-24    0.019508
2016-03-25    0.018598
2016-03-26    0.016423
2016-03-27    0.015447
2016-03-28    0.020529
2016-03-29    0.021372
2016-03-30    0.024169
2016-03-31    0.023459
2016-04-01    0.022882
2016-04-02    0.024879
2016-04-03    0.024946
2016-04-04    0.024324
2016-04-05    0.126481
2016-04-06    0.225354
2016-04-07    0.134138
Name: last_seen, dtype: float64

In [18]:
autos_trim["registration_year"].describe()

count    45058.000000
mean      2005.063918
std         89.689852
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [19]:
autos_trim["registration_year"].unique()

array([2004, 1997, 2009, 2007, 2003, 2006, 1998, 2017, 2010, 1999, 1982,
       1990, 2015, 2014, 1992, 2002, 2000, 2012, 2011, 1995, 2005, 2008,
       1985, 2016, 1996, 1994, 1986, 2001, 2018, 2013, 1972, 1993, 1988,
       1989, 1973, 1967, 1976, 4500, 1987, 1991, 1983, 1960, 1969, 1978,
       1980, 1984, 1963, 1977, 1961, 1968, 1934, 1965, 1971, 1966, 1979,
       1981, 1970, 1974, 1975, 4100, 1956, 9999, 6200, 1964, 1959, 1958,
       1948, 1931, 1943, 1941, 1950, 1962, 1927, 1937, 1929, 1000, 1910,
       5000, 1957, 1952, 1955, 1939, 8888, 1954, 1938, 2800, 5911, 1953,
       1951, 4800, 2019, 1001, 9000])

In [8]:
autos_trim["registration_year"].value_counts().sort_index(ascending=True)

1000    1
1001    1
1910    2
1927    1
1929    1
       ..
5911    1
6200    1
8888    1
9000    1
9999    3
Name: registration_year, Length: 93, dtype: int64

In the ```registration_year``` column there are values which are strange. Years like 1000 or 1001 are impossible,since the monks didn't ride automobiles back then. Also anything above 20xx is also an anomaly. Besides, since this dataset is from 2016, anything above 2016 is inaccurate.

In [9]:
autos_trim.loc[autos_trim["registration_year"]>2000,
               "registration_year"].value_counts().sort_index(ascending=True)

2001    2519
2002    2419
2003    2657
2004    2683
2005    2864
2006    2666
2007    2269
2008    2204
2009    2078
2010    1582
2011    1616
2012    1304
2013     793
2014     650
2015     360
2016     944
2017    1296
2018     462
2019       1
2800       1
4100       1
4500       1
4800       1
5000       2
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

Exploring the cars that fall inside the 1900-2016 boundary.

In [10]:
autos_trim[autos_trim["registration_year"].between(1900, 2016)].shape

(43284, 20)

In [11]:
autos_trim[~(autos_trim["registration_year"].between(1900, 2016))].shape

(1774, 20)

In [12]:
autos_trim.loc[autos_trim["registration_year"].between(1900,1950),
               "registration_year"].value_counts().sort_index(ascending=True)

1910    2
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    1
Name: registration_year, dtype: int64

In [33]:
autos_trim.loc[autos_trim["registration_year"].between(1900,1950),
               "registration_year"].value_counts().sort_index(ascending=True).index

Int64Index([1910, 1927, 1929, 1931, 1934, 1937, 1938, 1939, 1941, 1943, 1948,
            1950],
           dtype='int64')

In [13]:
years = [1910, 1927, 1929, 1931, 1934, 1937, 1938, 1939, 1941, 1943, 1948,
            1950]

In [14]:
autos_trim[autos_trim["registration_year"] == 1910]

date_crawled            name  seller offer_type  price_in_$  \
22659  2016-03-14 08:51:18    Opel_Corsa_B  privat    Angebot       500.0   
28693  2016-03-22 17:48:41  Renault_Twingo  privat    Angebot       599.0   

       ab_test vehicle_type  registration_year  gearbox  power_PS  model  \
22659     test          NaN               1910      NaN         0  corsa   
28693  control   kleinwagen               1910  manuell         0    NaN   

       odometer_in_km  registration_month fuel_type    brand  \
22659        150000.0                   0       NaN     opel   
28693          5000.0                   0    benzin  renault   

      unrepaired_damage           ad_created  num_of_pictures  postal_code  \
22659               NaN  2016-03-14 00:00:00                0        52393   
28693               NaN  2016-03-22 00:00:00                0        70376   

                 last_seen  
22659  2016-04-03 07:53:55  
28693  2016-04-06 09:16:59

1910 definitely can't be a year for those two models.

In [17]:
autos_trim.loc[autos_trim["registration_year"].isin(years),
               ["registration_year", "name", "model"]].sort_values("registration_year")

registration_year                                               name  \
22659               1910                                       Opel_Corsa_B   
28693               1910                                     Renault_Twingo   
21416               1927                            Essex_super_six__Ford_A   
22101               1929                                         BMW_Andere   
11246               1931                  Ford_Model_A_Roadster_Deluxe_1931   
2221                1934  Sehr_seltener_Oldtimer_Opel_1210_zum_Restaurieren   
2573                1934              Hanomag_rekord_15k_Suche_ersatz_teile   
26607               1937                                          MG_Andere   
23804               1937                                 Mercedes_Benz_170V   
21421               1937            Ford_Business_Coupe_Hotrod_Projekt.1937   
39725               1937  Dodge_1937_Hotrod_Flathead_Scheunenfund_Oldtim...   
26103               1938               Oldtimer_Sport___Cabrio_Georges_Irat   
24855               1939                      Andere_DKW_F8_Cabrio_Limosine   
13963               1941  Mercedes_Benz_L1500S_Wehrmacht_/_Luftwaffe___F...   
25792               1941              Andere_GAZ_Oldtimergelaendewagen_1941   
11585               1943                              Volkswagen__VW_Typ_82   
11047               1948          Andere_Simca_5_Fourgonette_Kombilimousine   
14020               1950                                 Oldtimeraufloesung   

        model  
22659   corsa  
28693     NaN  
21416  andere  
22101  andere  
11246  andere  
2221   andere  
2573      NaN  
26607     NaN  
23804  andere  
21421  andere  
39725     NaN  
26103     NaN  
24855     NaN  
13963  andere  
25792     NaN  
11585  andere  
11047     NaN  
14020  andere

In [18]:
autos_trim.loc[autos_trim["registration_year"].between(2000,2016),
               "registration_year"].value_counts().sort_index(ascending=True)

2000    2714
2001    2519
2002    2419
2003    2657
2004    2683
2005    2864
2006    2666
2007    2269
2008    2204
2009    2078
2010    1582
2011    1616
2012    1304
2013     793
2014     650
2015     360
2016     944
Name: registration_year, dtype: int64

Selecting 1927 as the starting year and 2016 as the end year.

In [19]:
autos_trim_year = (autos_trim[autos_trim["registration_year"]
                         .between(1927,2016)])

In [20]:
autos_trim_year.shape

(43282, 20)

In [21]:
autos_trim.shape

(45058, 20)

In [22]:
autos_trim_year["registration_year"].isnull().sum()

0

In [24]:
autos_trim_year["registration_year"].value_counts().sort_index()

1927       1
1929       1
1931       1
1934       2
1937       4
        ... 
2012    1304
2013     793
2014     650
2015     360
2016     944
Name: registration_year, Length: 77, dtype: int64

The ```brand``` column is up next.

In [25]:
autos_trim_year["brand"].describe()

count          43282
unique            40
top       volkswagen
freq            9180
Name: brand, dtype: object

In [26]:
autos_trim_year["brand"].isnull().sum()

0

In [27]:
autos_trim_year["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'renault', 'audi', 'sonstige_autos', 'mazda', 'porsche', 'mini',
       'mercedes_benz', 'seat', 'toyota', 'opel', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'jaguar', 'fiat', 'skoda', 'subaru',
       'kia', 'citroen', 'mitsubishi', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [30]:
autos_trim_year["brand"].value_counts().sort_values(ascending=False)

volkswagen        9180
bmw               4980
mercedes_benz     4427
opel              4346
audi              3920
ford              2822
renault           1901
peugeot           1272
fiat              1019
seat               775
skoda              743
smart              655
mazda              650
nissan             650
citroen            613
toyota             581
hyundai            440
sonstige_autos     415
volvo              406
mini               406
honda              338
mitsubishi         335
kia                318
alfa_romeo         287
chevrolet          261
suzuki             256
porsche            252
chrysler           156
dacia              123
jeep               106
land_rover          98
daihatsu            94
subaru              86
saab                72
jaguar              70
daewoo              58
rover               54
trabant             48
lancia              43
lada                26
Name: brand, dtype: int64

There are 40 unique brands of cars in the dataset with 'Volkswagen' being the runaway leader in the cars listed. I'm going to choose the top 15 brands to calculate some aggregate statistics.

In [32]:
brand_indices = autos_trim_year["brand"].value_counts().sort_values(ascending=False).index[:16]

In [33]:
'volkswagen' in brand_indices

True

In [40]:
brands_agg = {}
for brand in brand_indices:
    brands_agg[brand] = (autos_trim_year.loc[
        autos_trim_year["brand"] == brand,
        "price_in_$"].mean())
    
brands_agg

{'volkswagen': 5783.622984749455,
 'bmw': 8447.069879518072,
 'mercedes_benz': 8666.677208041563,
 'opel': 3394.705476300046,
 'audi': 9571.457397959184,
 'ford': 4247.120481927711,
 'renault': 2820.2272488164126,
 'peugeot': 3360.9205974842766,
 'fiat': 3256.152109911678,
 'seat': 4810.883870967742,
 'skoda': 6558.121130551817,
 'smart': 3611.0442748091605,
 'mazda': 4459.2692307692305,
 'nissan': 5175.8969230769235,
 'citroen': 4012.8336052202285,
 'toyota': 5266.97074010327}

In [42]:
dict(sorted(brands_agg.items(), key=lambda item: item[1], reverse=True))

{'audi': 9571.457397959184,
 'mercedes_benz': 8666.677208041563,
 'bmw': 8447.069879518072,
 'skoda': 6558.121130551817,
 'volkswagen': 5783.622984749455,
 'toyota': 5266.97074010327,
 'nissan': 5175.8969230769235,
 'seat': 4810.883870967742,
 'mazda': 4459.2692307692305,
 'ford': 4247.120481927711,
 'citroen': 4012.8336052202285,
 'smart': 3611.0442748091605,
 'opel': 3394.705476300046,
 'peugeot': 3360.9205974842766,
 'fiat': 3256.152109911678,
 'renault': 2820.2272488164126}

Choosing the top 15 brands of listed cars and aggregating the mean prices of each brand gives us 'Audi' to be the brand with cars having the highest mean price. Although 'Volkswagen' is the most listed car brand, it comes 4th in the list - more affordable than an 'Audi'. The top three brands are more expensive than the rest.

Now I'm exploring the prices and the mileage of the top brands to decipher if there's any connection, why the top brands are more expensive than the others.

In [43]:
brands_prices = pd.Series(brands_agg)
brands_prices

volkswagen       5783.622985
bmw              8447.069880
mercedes_benz    8666.677208
opel             3394.705476
audi             9571.457398
ford             4247.120482
renault          2820.227249
peugeot          3360.920597
fiat             3256.152110
seat             4810.883871
skoda            6558.121131
smart            3611.044275
mazda            4459.269231
nissan           5175.896923
citroen          4012.833605
toyota           5266.970740
dtype: float64

In [45]:
brands_mileage={}
for brand in brand_indices:
    brands_mileage[brand] = (autos_trim_year.loc[
        autos_trim_year["brand"] == brand,
        "odometer_in_km"].mean())
    
brands_mileage

{'volkswagen': 128234.74945533769,
 'bmw': 132928.71485943775,
 'mercedes_benz': 131083.12627061215,
 'opel': 128007.36309249885,
 'audi': 128941.32653061225,
 'ford': 123520.55279943303,
 'renault': 126415.04471330879,
 'peugeot': 126073.11320754717,
 'fiat': 114416.09421000982,
 'seat': 120058.06451612903,
 'skoda': 110861.3728129206,
 'smart': 99595.41984732824,
 'mazda': 123784.61538461539,
 'nissan': 116876.92307692308,
 'citroen': 118499.18433931484,
 'toyota': 115550.77452667814}

In [46]:
brands_mileage = pd.Series(brands_mileage)

In [47]:
prices_mileage = pd.DataFrame(brands_prices, columns = ["prices"])
prices_mileage["mileage"] = brands_mileage
prices_mileage

prices        mileage
volkswagen     5783.622985  128234.749455
bmw            8447.069880  132928.714859
mercedes_benz  8666.677208  131083.126271
opel           3394.705476  128007.363092
audi           9571.457398  128941.326531
ford           4247.120482  123520.552799
renault        2820.227249  126415.044713
peugeot        3360.920597  126073.113208
fiat           3256.152110  114416.094210
seat           4810.883871  120058.064516
skoda          6558.121131  110861.372813
smart          3611.044275   99595.419847
mazda          4459.269231  123784.615385
nissan         5175.896923  116876.923077
citroen        4012.833605  118499.184339
toyota         5266.970740  115550.774527

In [49]:
prices_mileage.sort_values("prices",ascending=False)

prices        mileage
audi           9571.457398  128941.326531
mercedes_benz  8666.677208  131083.126271
bmw            8447.069880  132928.714859
skoda          6558.121131  110861.372813
volkswagen     5783.622985  128234.749455
toyota         5266.970740  115550.774527
nissan         5175.896923  116876.923077
seat           4810.883871  120058.064516
mazda          4459.269231  123784.615385
ford           4247.120482  123520.552799
citroen        4012.833605  118499.184339
smart          3611.044275   99595.419847
opel           3394.705476  128007.363092
peugeot        3360.920597  126073.113208
fiat           3256.152110  114416.094210
renault        2820.227249  126415.044713

The data on the mean mileage clocked by the cars of each brand are not unambiguous to say that if prices are directly influenced by the mileage.